src/node.rs

```diff
@@ -662,40 +721,109 @@ impl EthNamespaceT for InMemoryNode {
+                    U64::from(1)
+                } else {
+                    U64::from(0)
+                }),
                 effective_gas_price: Some(500.into()),
-            }
-        });
+                ..Default::default()
+            });
 
-        Ok(receipt).into_boxed_future()
+            Ok(receipt).map_err(|_: jsonrpc_core::Error| into_jsrpc_error(Web3Error::InternalError))
+        })
     }
-
+    /// Sends a raw transaction to the L2 network.
+    ///
+    /// # Arguments
+    ///
+    /// * `tx_bytes` - The transaction bytes to send.
+    ///
+    /// # Returns
+    ///
+    /// A future that resolves to the hash of the transaction if successful, or an error if the transaction is invalid or execution fails.
     fn send_raw_transaction(
         &self,
         tx_bytes: zksync_basic_types::Bytes,
     ) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<zksync_basic_types::H256>> {
-        let chain_id = {
-            let reader = self.inner.read().unwrap();
-            reader.fork_storage.chain_id
+        let chain_id = match self.inner.read() {
+            Ok(reader) => reader.fork_storage.chain_id,
+            Err(_) => {
+                return futures::future::err(into_jsrpc_error(Web3Error::InternalError)).boxed()
+            }
         };
 
         let (tx_req, hash) =
-            TransactionRequest::from_bytes(&tx_bytes.0, chain_id.0, MAX_TX_SIZE).unwrap();
+            match TransactionRequest::from_bytes(&tx_bytes.0, chain_id.0, MAX_TX_SIZE) {
+                Ok(result) => result,
+                Err(e) => {
+                    return futures::future::err(into_jsrpc_error(Web3Error::SerializationError(e)))
+                        .boxed()
+                }
+            };
+
+        let mut l2_tx: L2Tx = match tx_req.try_into() {
+            Ok(tx) => tx,
+            Err(e) => {
+                return futures::future::err(into_jsrpc_error(Web3Error::SerializationError(e)))
+                    .boxed()
+            }
+        };
 
-        let mut l2_tx: L2Tx = tx_req.try_into().unwrap();
         l2_tx.set_input(tx_bytes.0, hash);
-        assert_eq!(hash, l2_tx.hash());
+        if hash != l2_tx.hash() {
+            return futures::future::err(into_jsrpc_error(Web3Error::InvalidTransactionData(
+                zksync_types::ethabi::Error::InvalidData,
+            )))
+            .boxed();
+        };
 
-        self.run_l2_tx(l2_tx, TxExecutionMode::VerifyExecute);
+        match self.run_l2_tx(l2_tx.clone(), TxExecutionMode::VerifyExecute) {
+            Ok(_) => Ok(hash).into_boxed_future(),
+            Err(e) => {
+                let error_message = format!("Execution error: {}", e);
+                futures::future::err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                    error_message,
+                    l2_tx.hash().as_bytes().to_vec(),
+                )))
+                .boxed()
+            }
+        };
 
         Ok(hash).into_boxed_future()
```

Comment:
Based on the `match` above, is this still necessary?

Reply:
ahh nice catch 👁️